Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,600 @@
+
+    /// The lock used by the connection pool used to ensure correct synchronization of accesses to `_connectionProviders`
+    ///
+    /// Accesses that only need synchronization for the span of a get or set can use the subscript instead of this property
+    ///
+    /// - Warning: This lock should always be acquired *before* `HTTP1ConnectionProvider`s `stateLock` if used in combination with it.
+    private let connectionProvidersLock = Lock()
+
+    init(configuration: HTTPClient.Configuration) {
+        self.configuration = configuration
+    }
+
+    /// Gets the `EventLoop` associated to the given `Key` if it exists
+    ///
+    /// This is part of optimization used by the `.execute(...)` method when
+    /// a request has its `EventLoopPreference` property set to `.indifferent`.
+    /// Having a default `EventLoop` shared by the *channel* and the *delegate* avoids
+    /// loss of performance due to `EventLoop` hoping
+    func associatedEventLoop(for key: Key) -> EventLoop? {
+        return self.connectionProvidersLock.withLock {
+            self.connectionProviders[key]?.eventLoop
+        }
+    }
+
+    /// This method asks the pool for a connection usable by the specified `request`, respecting the specified options.
+    ///
+    /// - parameter request: The request that needs a `Connection`
+    /// - parameter preference: The `EventLoopPreference` the connection pool will respect to lease a new connection
+    /// - parameter deadline: The connection timeout
+    /// - Returns: A connection  corresponding to the specified parameters
+    ///
+    /// When the pool is asked for a new connection, it creates a `Key` from the url associated to the `request`. This key
+    /// is used to determine if there already exists an associated `HTTP1ConnectionProvider` in `connectionProviders`
+    /// if it does, the connection provider then takes care of leasing a new connection. If a connection provider doesn't exist, it is created.
+    func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, on eventLoop: EventLoop, deadline: NIODeadline?) -> EventLoopFuture<Connection> {
+        let key = Key(request)
+
+        let provider: HTTP1ConnectionProvider = self.connectionProvidersLock.withLock {
+            if let existing = self.connectionProviders[key] {
+                existing.stateLock.withLock {
+                    existing.state.pending += 1
+                }
+                return existing
+            } else {
+                let http1Provider = HTTP1ConnectionProvider(key: key, eventLoop: eventLoop, configuration: self.configuration, parentPool: self)
+                self.connectionProviders[key] = http1Provider
+                http1Provider.stateLock.withLock {
+                    http1Provider.state.pending += 1
+                }
+                return http1Provider
+            }
+        }
+
+        return provider.getConnection(preference: preference)
```

Comment:
@adtrevor I think there's a race here:
- 1: line 81 (this file): we released the lock
- 2: line 83 (this file): then we call `provider.getConnection`
- 3: `provider.getConnection` (under the _re-acquired_ lock) which calls `connectionAction`
- 4: `connectionAction` will then remove the connection from the available connections

The problem is that between (1) and (2)/(3) we might get another `getConnection` call which then might find the _same_ connection `provider`

Reply:
I'd hope that https://github.com/swift-server/async-http-client/pull/160 finds some of these issues

Reply:
As we discussed I think this is actually ok, discovered new synchronisation issue but it is not related to this. Merged #160.